src/cache.rs

```diff
@@ -144,20 +170,39 @@ impl Cache {
 
     /// Check for pages for a given platform in one of the given languages.
     fn find_page_for_platform(
-        name: &str,
+        page_name: &str,
         cache_dir: &Path,
         platform: &str,
         language_dirs: &[String],
-    ) -> Option<PathBuf> {
+    ) -> Option<PageLookupResult> {
         language_dirs
             .iter()
-            .map(|lang_dir| cache_dir.join(lang_dir).join(platform).join(name))
+            .map(|lang_dir| cache_dir.join(lang_dir).join(platform).join(page_name))
             .find(|path| path.exists() && path.is_file())
+            .map(|file| PageLookupResult::with_page(file))
+    }
+
+    /// Look up custom patch (<name>.patch). If it exists, store it in a variable.
+    fn find_patch(patch_name: &str, maybe_custom_dir: Option<&Path>) -> Option<PathBuf> {
+        if let Some(custom_dir) = maybe_custom_dir {
+            let patch_path = custom_dir.join(patch_name);
+            if patch_path.exists() && patch_path.is_file() {
+                return Some(patch_path);
+            }
+        }
+        None
```

Comment:
This function can be converted with some Rust magic:
```rust
maybe_custom_dir
    .map(|custom_dir| custom_dir.join(patch_name))
    .filter(|path| path.exists() && path.is_file())
```

Reply:
filter on an option. I wouldn't have known that. Super cool.